pkg/controllers/disruption/emptiness.go

```diff
@@ -18,74 +18,120 @@ package disruption
+func (e *Emptiness) ComputeCommand(ctx context.Context, disruptionBudgetMapping map[string]map[v1.DisruptionReason]int, candidates ...*Candidate) (Command, scheduling.Results, error) {
+	if e.IsConsolidated() {
+		return Command{}, scheduling.Results{}, nil
+	}
+	candidates = e.sortCandidates(candidates)
+
+	empty := make([]*Candidate, 0, len(candidates))
+	constrainedByBudgets := false
+	for _, candidate := range candidates {
+		if len(candidate.reschedulablePods) > 0 {
+			continue
+		}
+		if disruptionBudgetMapping[candidate.nodePool.Name][v1.DisruptionReasonEmpty] == 0 {
+			// set constrainedByBudgets to true if any node was a candidate but was constrained by a budget
+			constrainedByBudgets = true
+			continue
+		}
+		// If there's disruptions allowed for the candidate's nodepool,
+		// add it to the list of candidates, and decrement the budget.
+		empty = append(empty, candidate)
+		disruptionBudgetMapping[candidate.nodePool.Name][v1.DisruptionReasonEmpty]--
+	}
+	// none empty, so do nothing
+	if len(empty) == 0 {
+		// if there are no candidates, but a nodepool had a fully blocking budget,
+		// don't mark the cluster as consolidated, as it's possible this nodepool
+		// should be consolidated the next time we try to disrupt.
+		if !constrainedByBudgets {
+			e.markConsolidated()
+		}
+		return Command{}, scheduling.Results{}, nil
+	}
+
+	cmd := Command{
+		candidates: empty,
+	}
+
+	// Empty Node Consolidation doesn't use Validation as we get to take advantage of cluster.IsNodeNominated.  This
+	// lets us avoid a scheduling simulation (which is performed periodically while pending pods exist and drives
+	// cluster.IsNodeNominated already).
+	select {
+	case <-ctx.Done():
+		return Command{}, scheduling.Results{}, errors.New("interrupted")
+	case <-e.clock.After(consolidationTTL):
+	}
+
+	v := NewValidation(e.clock, e.cluster, e.kubeClient, e.provisioner, e.cloudProvider, e.recorder, e.queue, v1.DisruptionReasonEmpty)
+	validatedCandidates, err := v.ValidateCandidates(ctx, cmd.candidates...)
+	if err != nil {
+		if IsValidationError(err) {
+			log.FromContext(ctx).V(1).Info(fmt.Sprintf("abandoning empty node consolidation attempt due to pod churn, command is no longer valid, %s", cmd))
+			return Command{}, scheduling.Results{}, nil
+		}
+		return Command{}, scheduling.Results{}, err
+	}
+
+	// TODO (jmdeal@): better encapsulate within validation
+	if lo.ContainsBy(validatedCandidates, func(c *Candidate) bool {
+		return len(c.reschedulablePods) != 0
+	}) {
+		log.FromContext(ctx).V(1).Info(fmt.Sprintf("abandoning empty node consolidation attempt due to pod churn, command is no longer valid, %s", cmd))
+		return Command{}, scheduling.Results{}, nil
+	}
+
+	return cmd, scheduling.Results{}, nil
 }
 
 func (e *Emptiness) Type() string {
-	return metrics.EmptinessReason
+	return metrics.ConsolidationReason
 }
 
 func (e *Emptiness) Class() string {
 	return GracefulDisruptionClass
 }
 
 func (e *Emptiness) ConsolidationType() string {
```

Comment:
It's pretty much just because of the timeouts that we have. I can remove it in favor of hard-coded values to simplify the interface, but the metric + label would still exist

Reply:
I've definitely seen people reference it. We should minimally keep the metric, but keeping the label has value, as it's a core part of how Karpenter works, especially at v1. The only way this changes is if we merge them in the future, which I don't know how that will be implemented, if at all.

Reply:
Sounds good. Discussed offline. Seems like we are fine to keep this for now